LogicalTypes.md

```diff
@@ -148,3 +148,179 @@ primitive type. The `binary` data is interpreted as an encoded BSON document as
+  single field named `array`. The repetition of this level must be either
+  `optional` or `required` and determines whether the list is nullable.
+* The middle level, named `array`, must be a repeated group with a single
+  field named `element`.
+* The `element` field encodes the list's element type and repetition.
+
+The following examples demonstrate two of the possible lists of string values.
+
+```
+// List<String> (list non-null, elements nullable)
+required group my_list (LIST) {
+  repeated group array {
+    optional binary element (UTF8);
+  }
+}
+
+// List<String> (list nullable, elements non-null)
+optional group my_list (LIST) {
+  repeated group array {
+    required binary element (UTF8);
+  }
+}
+```
+
+Element types can be nested structures. For example, a list of lists:
+
+```
+// List<List<Integer>>
+optional group array_of_arrays (LIST) {
+  repeated group array {
+    required group element (LIST) {
+      repeated group array {
+        required int32 element;
+      }
+    }
+  }
+}
+```
+
+It is *recommended* that the repeated group of elements is named `array` and
+that its element field is named `element`. However, these names should not be
+enforced as errors.
+
+#### Backward-compatibility rules
+
+Some existing data uses `LIST` to annotate the repeated group named `array` in
+the above requirements, without the outer group. For backward-compatibility,
+this structure must be supported as a required list named for the repeated group.
+
+```
+// List<String> (non-null list)
+repeated group my_list (LIST) {
+  required binary element (UTF8);
+}
+```
+
+Some existing data did not include the inner element layer. For
+backward-compatibility, the type of elements in `LIST`-annotated structures
+should always be determined by the following rules based on the repeated field:
+
+1. If the repeated field is not a group, then its type is the element type and
+   elements are required.
+2. If the repeated field is a group with multiple fields, then its type is the
+   element type and elements are required.
+3. If the repeated field is a group with one field, then the field's type is
+   the element type with the field's repetition.
+
+Examples that can be interpreted using these rules:
+
+```
+// List<Integer> (non-null list, nullable elements)
+repeated group my_list (LIST) {
+  optional int32 element;
+}
+
+// List<Integer> (nullable list, non-null elements)
+optional group my_list (LIST) {
+  repeated int32 element;
+}
+
+// List<Tuple<String, Integer>> (nullable list, non-null elements)
+optional group my_list (LIST) {
+  repeated group element {
+    required binary str (UTF8);
+    required int32 num;
+  };
+}
+```
+
+### Maps
+
+`MAP` is used to annotate types that should be interpreted as a map from keys
+to values. `MAP` must annotate a 3-level structure:
+
+```
+<map-repetition> group <name> (MAP) {
+  repeated group key_value {
+    required <key-type> key;
+    <value-repetition> <value-type> value;
```

Comment:
I don't think we would mix using and not using List annotations in the same schema. Either we use repeated without annotation or we don't. We should not mix those.
The consequence is this element can be only required or optional.